### PR TITLE
feat(crypto): CRP-2579 Add support for derivation to ecdsa_secp256r1 crate

### PR DESCRIPTION
**Pull request by randombit**
In order to remove the dependency on the internal threshold ECDSA protocol implementation from ic-crypto-utils-canister-threshold-sig we have to have derivation available in the 3 signature utility crates.

### CODE REVIEW

File: rs/crypto/ecdsa_secp256r1/src/lib.rs [hunk 1]
```diff
@@ -305,6 +434,57 @@ impl PrivateKey {
...
+        let (_pt, offset, derived_chain_code) =
+            derivation_path.derive_offset(public_key, chain_code);
+
+        let derived_scalar = self.key.as_nonzero_scalar().as_ref().add(&offset);
+
+        let nz_ds =
+            NonZeroScalar::new(derived_scalar).expect("Derivation always produces non-zero sum");
+
+        let derived_key = Self {
+            key: p256::SecretKey::from(nz_ds).into(),
```
Comment (fspreiss) [inline rs/crypto/ecdsa_secp256r1/src/lib.rs:483]: It seems we can also go directly: `p256::ecdsa::SigningKey::from(nz_ds)`

File: rs/crypto/ecdsa_secp256r1/src/lib.rs [hunk 2]
```diff
@@ -374,7 +554,7 @@ impl PublicKey {
     /// valid signature, then (r,-s) is also valid. To avoid this malleability,
     /// some systems require that s be "normalized" to the smallest value.
     ///
-    /// This normalization is quite common on secp256k1, but is virtually
+    /// This normalization is quite common on secp256r1, but is virtually
```
Comment (fspreiss) [inline rs/crypto/ecdsa_secp256r1/src/lib.rs:557]: Was this change intentional? The text now says: "normalization is common on r1, but is unknown for r1".

File: rs/crypto/ecdsa_secp256r1/src/lib.rs [hunk 3]
```diff
@@ -399,4 +579,39 @@ impl PublicKey {
...
+        chain_code: &[u8; 32],
+    ) -> (Self, [u8; 32]) {
+        let public_key: AffinePoint = *self.key.as_affine();
+        let (pt, _offset, chain_code) = derivation_path.derive_offset(public_key, chain_code);
+
+        let derived_key = Self {
+            key: p256::PublicKey::from(
+                p256::PublicKey::from_affine(pt).expect("Derived point is valid"),
+            )
+            .into(),
```
Comment (fspreiss) [inline rs/crypto/ecdsa_secp256r1/src/lib.rs:612]: It seems one conversion layer is to itself, so it should be obsolete. Also, how about being explicit (instead of doing `into`). As a result, I suggest we do
```
            key: p256::ecdsa::VerifyingKey::from(
                p256::PublicKey::from_affine(pt).expect("Derived point is valid"),
            ),
```

File: rs/crypto/ecdsa_secp256r1/tests/tests.rs [hunk 4]
```diff
@@ -271,3 +272,94 @@ NRLvCGaIxJfchxpjcCysTG12MfKOf6/Phw==
...
+
+        assert_eq!(hex::encode(cc_pk), hex::encode(cc_sk));
+
+        let msg = rng.gen::<[u8; 32]>();
+        let derived_sig = derived_sk.sign_message(&msg);
+
+        assert!(derived_pk.verify_signature(&msg, &derived_sig));
+    }
+}
+
```
Comment (fspreiss) [inline rs/crypto/ecdsa_secp256r1/tests/tests.rs:313]: For some key derivations, we also have a test `private_derivation_also_works_for_derived_keys`: would that be applicable here as well?

**Review by fspreiss — COMMENTED** [review]
Thanks, @randombit! Looks good to me! Find some minor comments below.

File: rs/crypto/ecdsa_secp256r1/src/lib.rs [hunk 5]
```diff
@@ -305,6 +434,57 @@ impl PrivateKey {
...
+    /// Derive a private key from this private key using a derivation path
+    ///
+    /// This is the same derivation system used by the Internet Computer when
+    /// deriving subkeys for threshold ECDSA with secp256r1
+    ///
+    /// As long as each index of the derivation path is a 4-byte input with the highest
+    /// bit cleared, this derivation scheme matches SLIP-10
+    ///
+    /// See <https://internetcomputer.org/docs/current/references/ic-interface-spec#ic-ecdsa_public_key>
+    /// for details on the derivation scheme.
```
[the remark this answers is not in the document]
Comment (randombit) [reply rs/crypto/ecdsa_secp256r1/src/lib.rs:447]: Opened https://dfinity.atlassian.net/browse/CRP-2627

File: rs/crypto/ecdsa_secp256r1/src/lib.rs (hunk 2 above)
Comment (randombit) [reply rs/crypto/ecdsa_secp256r1/src/lib.rs:557]: Oops search and replace gone bad

File: rs/crypto/ecdsa_secp256r1/tests/tests.rs (hunk 4 above)
Comment (randombit) [reply rs/crypto/ecdsa_secp256r1/tests/tests.rs:313]: That test was written for and primarily relevant for Ed25519 because there (due to how Ed25519 secret keys are defined, as a seed that is hashed to produce the key), you can't perform an additive key derivation with a standard Ed25519 key. Instead you have to use the expanded key, which is the scalar + derandomization secret which are derived from the private seed by hashing it with SHA-512. So the return value of `derive_key` on the Ed25519 private key returns a completely different type, which cannot be serialized (since doing so would require computing a SHA-512 preimage) but can only sign. However it can (and does) support further derivation if required, and this test checks that derivation algorithms of `ed25519::PrivateKey` and `ed25519::DerivedPrivateKey` behave the same.

That said the same test does logically make sense here; for all derivation paths, chain codes, and keys, deriving the entire path at once is identical to deriving twice with the path split into two. So I imported the test.

File: rs/crypto/ecdsa_secp256r1/src/lib.rs (hunk 3 above)
Comment (randombit) [reply rs/crypto/ecdsa_secp256r1/src/lib.rs:612]: Done

File: rs/crypto/ecdsa_secp256r1/src/lib.rs (hunk 1 above)
Comment (randombit) [reply rs/crypto/ecdsa_secp256r1/src/lib.rs:483]: Fixed, thanks

File: rs/crypto/ecdsa_secp256r1/src/lib.rs [hunk 6]
```diff
@@ -305,6 +434,57 @@ impl PrivateKey {
...
+    ///
+    pub fn derive_subkey_with_chain_code(
+        &self,
+        derivation_path: &DerivationPath,
+        chain_code: &[u8; 32],
+    ) -> (Self, [u8; 32]) {
+        use p256::NonZeroScalar;
+
+        let public_key: AffinePoint = self.key.verifying_key().as_affine().clone();
```
[the remark this answers is not in the document]
Comment (randombit) [reply rs/crypto/ecdsa_secp256r1/src/lib.rs:473]: Fixed

File: rs/crypto/ecdsa_secp256r1/src/lib.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (randombit) [reply rs/crypto/ecdsa_secp256r1/src/lib.rs:450]: During discussion in the team meeting we realized updating the interface spec would not be appropriate since the IC itself does not currently support P-256 signatures. Instead this reference to the interface spec should be removed.

File: rs/crypto/ecdsa_secp256r1/src/lib.rs [hunk 7]
```diff
@@ -35,6 +35,135 @@ lazy_static::lazy_static! {
...
+///
+/// A derivation path is simply a sequence of DerivationIndex
+#[derive(Clone, Debug)]
+pub struct DerivationPath {
+    path: Vec<DerivationIndex>,
+}
+
+impl DerivationPath {
+    /// Create a BIP32-style derivation path
+    pub fn new_bip32(bip32: &[u32]) -> Self {
```
[the remark this answers is not in the document]
Comment (fspreiss) [reply rs/crypto/ecdsa_secp256r1/src/lib.rs]: Thanks for adding the reference to SLIP-10. No strong opinion: Both names are OK with me.

File: rs/crypto/ecdsa_secp256r1/tests/tests.rs (hunk 4 above)
Comment (fspreiss) [reply rs/crypto/ecdsa_secp256r1/tests/tests.rs:313]: 👍

**Review by fspreiss — APPROVED** [review]

File: rs/crypto/ecdsa_secp256r1/src/lib.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (randombit) [reply rs/crypto/ecdsa_secp256r1/src/lib.rs]: Removed refs to the interface spec for the time being
